WCSimfitQunHitPreprocess.c

```diff
@@ -0,0 +1,368 @@
+        WCSimEntrance[1] = wcsimroottrack->GetBoundaryPoints().at(1).at(2)/10;    //g
+        WCSimEntrance[2] = wcsimroottrack->GetBoundaryPoints().at(1).at(1)/10;    //g
+        WCSimExit[0] = -wcsimroottrack->GetBoundaryPoints().at(2).at(0)/10;    //g
+        WCSimExit[1] = wcsimroottrack->GetBoundaryPoints().at(2).at(2)/10;    //g
+        WCSimExit[2] = wcsimroottrack->GetBoundaryPoints().at(2).at(1)/10;    //g
+
+        //calculate the WCSim travel direction
+        WCSimDir[0] = WCSimExit[0] - WCSimEntrance[0];
+        WCSimDir[1] = WCSimExit[1] - WCSimEntrance[1];
+        WCSimDir[2] = WCSimExit[2] - WCSimEntrance[2];
+        double WCSimDist = sqrt(WCSimDir[0]*WCSimDir[0]+WCSimDir[1]*WCSimDir[1]+WCSimDir[2]*WCSimDir[2]);
+        WCSimDir[0] = WCSimDir[0]/WCSimDist;    //g
+        WCSimDir[1] = WCSimDir[1]/WCSimDist;    //g
+        WCSimDir[2] = WCSimDir[2]/WCSimDist;    //g
+        WCSimStartPos[0] = -wcsimroottrack->GetStart(0);    //g
+        WCSimStartPos[1] = wcsimroottrack->GetStart(2);    //g
+        WCSimStartPos[2] = wcsimroottrack->GetStart(1);    //g
+        THNum = wcsimrootevent->GetNcherenkovhittimes();
+
+        TClonesArray* WRCDHArray = wcsimrootevent->GetCherenkovDigiHits();    //Array of WCSimRootCherenkovDigiHit pointer
+        DHNum = WRCDHArray->GetEntriesFast();
+        for (int j=0; j<DHNum; j++) {
+            //ignore extremely long digit time (>200 ns)
+            if (((WCSimRootCherenkovDigiHit*) WRCDHArray->At(j))->GetT()>200) {
+                WRCDHArray->RemoveAt(j);
+            }
+        }
+        WRCDHArray->Compress();
+        DHNum = WRCDHArray->GetEntriesFast();    //g
+        //loop the array and fill the DHTree
+        for (int j=0; j<DHNum; j++) {
+            WCSimRootCherenkovDigiHit* WRCDH = (WCSimRootCherenkovDigiHit*) WRCDHArray->At(j);
+            DHPMT = WRCDH->GetTubeId()-1;   //g
+            DHTime = WRCDH->GetT();     //g
+            Charge = WRCDH->GetQ();     //g
+            DHTree->Fill();
+        }
+
+        //loop for the PMTs with true hit in this event
+        int THPMTNum = wcsimrootevent->GetNcherenkovhits();
+        for (int j=0; j<THPMTNum; j++) {
+            WCSimRootCherenkovHit* WRCH = (WCSimRootCherenkovHit*) wcsimrootevent->GetCherenkovHits()->At(j);
+            THPMT = WRCH->GetTubeID()-1;    //g
+
+            int FirstHitPos = WRCH->GetTotalPe(0), LastHitPos = WRCH->GetTotalPe(0)+WRCH->GetTotalPe(1)-1;     //The position on the Array of WCSimRootCherenkovHitTime of the first hit and the last hit on this PMT
+            //loop for the true hits of a PMT in an event and fill the THTree
+            for (int k=FirstHitPos; k<=LastHitPos; k++) {
+                WCSimRootCherenkovHitTime* WRCHT = (WCSimRootCherenkovHitTime*) wcsimrootevent->GetCherenkovHitTimes()->At(k);
+                if (WRCHT->GetParentSavedTrackID()==(-1)) {
+                    THNum--;
+                    continue;
+                }   //ignore dark noise
+
```

Comment:
We actually don't care whether a true hit generates a digi hit or not, so you don't need the loop below for selection. Just record every true hit.

Reply:
The loop below is not for selection of true hits but to find the DHID. That is, if a true hit generate a digit hit, the DHID of this true hit carries the position of corresponding digit hit in the digit hit tree. This provides a method to relates the true and digit hit tree, though I didn't use this a lot.